<a href="https://colab.research.google.com/github/BrianBehnke/DS-Unit-1-Sprint-4-Statistical-Tests-and-Experiments/blob/master/DS_Unit_1_Sprint_Challenge_4.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Data Science Unit 1 Sprint Challenge 4

## Exploring Data, Testing Hypotheses

In this sprint challenge you will look at a dataset of people being approved or rejected for credit.

https://archive.ics.uci.edu/ml/datasets/Credit+Approval

Data Set Information: This file concerns credit card applications. All attribute names and values have been changed to meaningless symbols to protect confidentiality of the data. This dataset is interesting because there is a good mix of attributes -- continuous, nominal with small numbers of values, and nominal with larger numbers of values. There are also a few missing values.

Attribute Information:
- A1: b, a.
- A2: continuous.
- A3: continuous.
- A4: u, y, l, t.
- A5: g, p, gg.
- A6: c, d, cc, i, j, k, m, r, q, w, x, e, aa, ff.
- A7: v, h, bb, j, n, z, dd, ff, o.
- A8: continuous.
- A9: t, f.
- A10: t, f.
- A11: continuous.
- A12: t, f.
- A13: g, p, s.
- A14: continuous.
- A15: continuous.
- A16: +,- (class attribute)

Yes, most of that doesn't mean anything. A16 (the class attribute) is the most interesting, as it separates the 307 approved cases from the 383 rejected cases. The remaining variables have been obfuscated for privacy - a challenge you may have to deal with in your data science career.

Sprint challenges are evaluated based on satisfactory completion of each part. It is suggested you work through it in order, getting each aspect reasonably working, before trying to deeply explore, iterate, or refine any given step. Once you get to the end, if you want to go back and improve things, go for it!

## Part 1 - Load and validate the data

- Load the data as a `pandas` data frame.
- Validate that it has the appropriate number of observations (you can check the raw file, and also read the dataset description from UCI).
- UCI says there should be missing data - check, and if necessary change the data so pandas recognizes it as na
- Make sure that the loaded features are of the types described above (continuous values should be treated as float), and correct as necessary

This is review, but skills that you'll use at the start of any data exploration. Further, you may have to do some investigation to figure out which file to load from - that is part of the puzzle.

In [0]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
from scipy import stats

url = 'https://archive.ics.uci.edu/ml/machine-learning-databases/credit-screening/crx.data'


In [11]:
# try read_csv
# maybe read_table may work better
names = ['A1', 'A2', 'A3', 'A4', 'A5', 'A6', 'A7', 'A8', 'A9', 'A10', 'A11', 'A12', 'A13', 'A14', 'A15', 'A16']
df = pd.read_csv(url, na_values='?', names=names)
df.head()

A1     A2     A3 A4 A5 A6 A7    A8 A9 A10  A11 A12 A13    A14  A15 A16
0  b  30.83  0.000  u  g  w  v  1.25  t   t    1   f   g  202.0    0   +
1  a  58.67  4.460  u  g  q  h  3.04  t   t    6   f   g   43.0  560   +
2  a  24.50  0.500  u  g  q  h  1.50  t   f    0   f   g  280.0  824   +
3  b  27.83  1.540  u  g  w  v  3.75  t   t    5   t   g  100.0    3   +
4  b  20.17  5.625  u  g  w  v  1.71  t   f    0   f   s  120.0    0   +

In [12]:
# expect 690 records, per data info page.
# Looks correct as we returned 690
df.describe()

A2          A3          A8        A11          A14  \
count  678.000000  690.000000  690.000000  690.00000   677.000000   
mean    31.568171    4.758725    2.223406    2.40000   184.014771   
std     11.957862    4.978163    3.346513    4.86294   173.806768   
min     13.750000    0.000000    0.000000    0.00000     0.000000   
25%     22.602500    1.000000    0.165000    0.00000    75.000000   
50%     28.460000    2.750000    1.000000    0.00000   160.000000   
75%     38.230000    7.207500    2.625000    3.00000   276.000000   
max     80.250000   28.000000   28.500000   67.00000  2000.000000   

                 A15  
count     690.000000  
mean     1017.385507  
std      5210.102598  
min         0.000000  
25%         0.000000  
50%         5.000000  
75%       395.500000  
max    100000.000000

In [19]:
# .shape also returns the correct number of records, 690
df.shape

(690, 16)

## Part 2 - Exploring data, Testing hypotheses

The only thing we really know about this data is that A16 is the class label. Besides that, we have 6 continuous (float) features and 9 categorical features.

Explore the data: you can use whatever approach (tables, utility functions, visualizations) to get an impression of the distributions and relationships of the variables. In general, your goal is to understand how the features are different when grouped by the two class labels (`+` and `-`).

For the 6 continuous features, how are they different when split between the two class labels? Choose two features to run t-tests (again split by class label) - specifically, select one feature that is *extremely* different between the classes, and another feature that is notably less different (though perhaps still "statistically significantly" different). You may have to explore more than two features to do this.

For the categorical features, explore by creating "cross tabs" between them and the class label, and apply the Chi-squared test to them. There are 9 categorical features - as with the t-test, try to find one where the Chi-squared test returns an extreme result (rejecting the null that the data are independent), and one where it is less extreme.

**NOTE** - "less extreme" just means smaller test statistic/larger p-value. Even the least extreme differences may be strongly statistically significant.

Your *main* goal is the hypothesis tests, so don't spend too much time on the exploration/visualization piece. That is just a means to an end. This is challenging, so manage your time and aim for a baseline of at least running two t-tests and two Chi-squared tests before polishing. And don't forget to answer the questions in part 3, even if your results in this part aren't what you want them to be.

In [13]:
# Split the data into 2 groups, based on the class attribute, A16, either + or -
good = df[df.A16 == '+']
bad = df[df.A16 == '-']

print(good.head())
print(bad.head())

  A1     A2     A3 A4 A5 A6 A7    A8 A9 A10  A11 A12 A13    A14  A15 A16
0  b  30.83  0.000  u  g  w  v  1.25  t   t    1   f   g  202.0    0   +
1  a  58.67  4.460  u  g  q  h  3.04  t   t    6   f   g   43.0  560   +
2  a  24.50  0.500  u  g  q  h  1.50  t   f    0   f   g  280.0  824   +
3  b  27.83  1.540  u  g  w  v  3.75  t   t    5   t   g  100.0    3   +
4  b  20.17  5.625  u  g  w  v  1.71  t   f    0   f   s  120.0    0   +
   A1     A2   A3 A4 A5  A6  A7      A8 A9 A10  A11 A12 A13    A14  A15 A16
70  b  32.33  7.5  u  g   e  bb   1.585  t   f    0   t   s  420.0    0   -
71  b  34.83  4.0  u  g   d  bb  12.500  t   f    0   t   g    NaN    0   -
72  a  38.58  5.0  u  g  cc   v  13.500  t   f    0   t   g  980.0    0   -
73  b  44.25  0.5  u  g   m   v  10.750  t   f    0   f   s  400.0    0   -
74  b  44.83  7.0  y  p   c   v   1.625  f   f    0   f   g  160.0    2   -


In [15]:
# quick glance at data from the good dataset
good.describe()

A2          A3          A8         A11         A14  \
count  305.000000  307.000000  307.000000  307.000000  301.000000   
mean    33.720492    5.904951    3.427899    4.605863  164.421927   
std     12.809641    5.471485    4.120792    6.320242  161.770675   
min     13.750000    0.000000    0.000000    0.000000    0.000000   
25%     23.170000    1.500000    0.750000    0.000000    0.000000   
50%     30.500000    4.460000    2.000000    3.000000  120.000000   
75%     41.330000    9.520000    5.000000    7.000000  280.000000   
max     76.750000   28.000000   28.500000   67.000000  840.000000   

                 A15  
count     307.000000  
mean     2038.859935  
std      7659.763941  
min         0.000000  
25%         0.000000  
50%       221.000000  
75%      1209.000000  
max    100000.000000

In [16]:
# quick glance at data from the bad dataset
bad.describe()

A2          A3          A8         A11          A14  \
count  373.000000  383.000000  383.000000  383.000000   376.000000   
mean    29.808231    3.839948    1.257924    0.631854   199.699468   
std     10.919291    4.337662    2.120481    1.900049   181.564835   
min     15.170000    0.000000    0.000000    0.000000     0.000000   
25%     22.000000    0.835000    0.125000    0.000000   100.000000   
50%     27.330000    2.210000    0.415000    0.000000   167.500000   
75%     34.830000    5.000000    1.500000    0.000000   272.000000   
max     80.250000   26.335000   13.875000   20.000000  2000.000000   

               A15  
count   383.000000  
mean    198.605744  
std     671.608839  
min       0.000000  
25%       0.000000  
50%       1.000000  
75%      67.000000  
max    5552.000000

In [135]:
df.describe()

A2          A3          A8        A11          A14  \
count  678.000000  690.000000  690.000000  690.00000   677.000000   
mean    31.568171    4.758725    2.223406    2.40000   184.014771   
std     11.957862    4.978163    3.346513    4.86294   173.806768   
min     13.750000    0.000000    0.000000    0.00000     0.000000   
25%     22.602500    1.000000    0.165000    0.00000    75.000000   
50%     28.460000    2.750000    1.000000    0.00000   160.000000   
75%     38.230000    7.207500    2.625000    3.00000   276.000000   
max     80.250000   28.000000   28.500000   67.00000  2000.000000   

                 A15  
count     690.000000  
mean     1017.385507  
std      5210.102598  
min         0.000000  
25%         0.000000  
50%         5.000000  
75%       395.500000  
max    100000.000000

## All columns look like they have higher averages and spreads in the good group, except for column A14, which is quite higher than in the bad group.

## Lets do some T TESTS! YAY!

In [136]:
# my two features I will look at are:
# large difference: A15
# small difference: A3

print('T Test for good.A15 ' , scipy.stats.ttest_1samp(good.A15, 1017.385))
print('T Test for bad.A15 ' , scipy.stats.ttest_1samp(bad.A15, 1017.385))

T Test for good.A15  Ttest_1sampResult(statistic=2.3365846338986787, pvalue=0.020105944183966846)
T Test for bad.A15  Ttest_1sampResult(statistic=-23.858866910185004, pvalue=1.1004621903621446e-77)


In [121]:
print('Two tailed T Test for good.A15 and bad.A15 ' , stats.ttest_ind(good.A15, bad.A15))

Two tailed T Test for good.A15 and bad.A15  Ttest_indResult(statistic=4.680216020964486, pvalue=3.4520256956287944e-06)


In [115]:
stats.bayes_mvs(good.A15, 0.05)

(Mean(statistic=2038.85993485342, minmax=(2011.4241811008205, 2066.2956886060197)),
 Variance(statistic=59057983.52295131, minmax=(58502684.883307904, 59099410.60717633)),
 Std_dev(statistic=7678.602001234962, minmax=(7648.704784687921, 7687.614103684988)))

In [116]:
stats.bayes_mvs(bad.A15, 0.05)

(Mean(statistic=198.60574412532637, minmax=(196.45238311545242, 200.75910513520031)),
 Variance(statistic=453432.424914113, minmax=(449801.28869636444, 453904.6340246738)),
 Std_dev(statistic=672.9310487206066, minmax=(670.6722662346823, 673.7244496266064)))

In [34]:
print('T Test for good.A3 ' , scipy.stats.ttest_1samp(good.A3, 0.5))
print('T Test for bad.A3 ' , scipy.stats.ttest_1samp(bad.A3, 0.5))

T Test for good.A3  Ttest_1sampResult(statistic=17.30835312313626, pvalue=2.8402779104335877e-47)
T Test for bad.A3  Ttest_1sampResult(statistic=15.068962886128801, pvalue=1.3340212686882321e-40)


In [117]:
stats.bayes_mvs(good.A3, 0.05)

(Mean(statistic=5.9049511400651475, minmax=(5.885353367038174, 5.924548913092121)),
 Variance(statistic=30.134103040352304, minmax=(29.850764100771357, 30.15524104660023)),
 Std_dev(statistic=5.4849413120426425, minmax=(5.463585278987723, 5.491378792853415)))

In [118]:
stats.bayes_mvs(bad.A3, 0.05)

(Mean(statistic=3.839947780678851, minmax=(3.8260400541579647, 3.853855507199737)),
 Variance(statistic=18.914338944620035, minmax=(18.762870859404046, 18.93403652838144)),
 Std_dev(statistic=4.346201565880141, minmax=(4.331612962789271, 4.351325835694385)))

In [120]:
print('Two tailed T Test for good.A3 and bad.A3 ' , stats.ttest_ind(good.A3, bad.A3))

Two tailed T Test for good.A3 and bad.A3  Ttest_indResult(statistic=5.52998337614816, pvalue=4.551680702308068e-08)


In [123]:
print('T Test for good.A2 ' , scipy.stats.ttest_1samp(good.A2, 0.5))
print('T Test for bad.A2 ' , scipy.stats.ttest_1samp(bad.A2, 0.5))
print('Two tailed T Test for good.A2 and bad.A2 ' , stats.ttest_ind(good.A2, bad.A2))

T Test for good.A2  Ttest_1sampResult(statistic=nan, pvalue=nan)
T Test for bad.A2  Ttest_1sampResult(statistic=nan, pvalue=nan)
Two tailed T Test for good.A2 and bad.A2  Ttest_indResult(statistic=nan, pvalue=nan)


In [124]:
print('T Test for good.A3 ' , scipy.stats.ttest_1samp(good.A3, 0.5))
print('T Test for bad.A3 ' , scipy.stats.ttest_1samp(bad.A3, 0.5))
print('Two tailed T Test for good.A3 and bad.A3 ' , stats.ttest_ind(good.A3, bad.A3))

T Test for good.A3  Ttest_1sampResult(statistic=17.30835312313626, pvalue=2.8402779104335877e-47)
T Test for bad.A3  Ttest_1sampResult(statistic=15.068962886128801, pvalue=1.3340212686882321e-40)
Two tailed T Test for good.A3 and bad.A3  Ttest_indResult(statistic=5.52998337614816, pvalue=4.551680702308068e-08)


In [125]:
print('T Test for good.A8 ' , scipy.stats.ttest_1samp(good.A8, 0.5))
print('T Test for bad.A8 ' , scipy.stats.ttest_1samp(bad.A8, 0.5))
print('Two tailed T Test for good.A8 and bad.A8 ' , stats.ttest_ind(good.A8, bad.A8))

T Test for good.A8  Ttest_1sampResult(statistic=12.449289487126055, pvalue=4.6132667541039515e-29)
T Test for bad.A8  Ttest_1sampResult(statistic=6.995051170689755, pvalue=1.1936831195754194e-11)
Two tailed T Test for good.A8 and bad.A8  Ttest_indResult(statistic=8.935819983773698, pvalue=3.6710537401601785e-18)


In [137]:
print('T Test for good.A11 ' , scipy.stats.ttest_1samp(good.A11, 2.400))
print('T Test for bad.A11 ' , scipy.stats.ttest_1samp(bad.A11, 2.400))

print('Two tailed T Test for good.A11 and bad.A11 ' , stats.ttest_ind(good.A11, bad.A11))

T Test for good.A11  Ttest_1sampResult(statistic=6.115247963474785, pvalue=2.9376851337382315e-09)
T Test for bad.A11  Ttest_1sampResult(statistic=-18.211793670382843, pvalue=8.54839203623047e-54)
Two tailed T Test for good.A11 and bad.A11  Ttest_indResult(statistic=11.667004222431277, pvalue=7.957718568079967e-29)


In [130]:
print('T Test for df.A11 ' , scipy.stats.ttest_1samp(df.A11, 0.5))
print('T Test for df.A15 ' , scipy.stats.ttest_1samp(df.A15, 0.5))

T Test for df.A11  Ttest_1sampResult(statistic=10.263115869755008, pvalue=4.298076068378268e-23)
T Test for df.A15  Ttest_1sampResult(statistic=5.126846652016273, pvalue=3.8324079891902166e-07)


In [127]:
print('T Test for good.A14 ' , scipy.stats.ttest_1samp(good.A14, 0.5))
print('T Test for bad.A14 ' , scipy.stats.ttest_1samp(bad.A14, 0.5))
print('Two tailed T Test for good.A14 and bad.A14 ' , stats.ttest_ind(good.A14, bad.A14))

T Test for good.A14  Ttest_1sampResult(statistic=nan, pvalue=nan)
T Test for bad.A14  Ttest_1sampResult(statistic=nan, pvalue=nan)
Two tailed T Test for good.A14 and bad.A14  Ttest_indResult(statistic=nan, pvalue=nan)


## Well, I still dont know if I understand this, so MORE tests! CHI SQUARED to the rescue!

So we will look at the chi squared of several categorical features of the dataset against A16, still seperated by good and bad.
We are using columns: A4, A1, and A6. I also did a chi quared on the entire dataset for columns A13 and A16.

In [72]:
good_tab_a4 = pd.crosstab(good.A4, good.A16)
print(good_tab_a4)

A16    +
A4      
l      2
u    256
y     45


In [100]:
from scipy.stats import chisquare
print('The Chi-Squared good results for A4 and A16 crosstab: ', chisquare(good_tab_a4, axis=None))

The Chi-Squared good results for A4 and A16 crosstab:  Power_divergenceResult(statistic=365.960396039604, pvalue=3.409649225976067e-80)


In [75]:
bad_tab_a4 = pd.crosstab(bad.A4, bad.A16)
print(bad_tab_a4)

A16    -
A4      
u    263
y    118


In [99]:
print('The Chi-Squared bad results for A4 and A16 crosstab: ', chisquare(bad_tab_a4, axis=None))

The Chi-Squared bad results for A4 and A16 crosstab:  Power_divergenceResult(statistic=55.183727034120736, pvalue=1.0977380519420046e-13)


In [77]:
good_tab_a1 = pd.crosstab(good.A1, good.A16)
print(good_tab_a1)

A16    +
A1      
a     98
b    206


In [98]:
print('The Chi-Squared good results for A1 and A16 crosstab: ', chisquare(good_tab_a1, axis=None))

The Chi-Squared good results for A1 and A16 crosstab:  Power_divergenceResult(statistic=38.36842105263158, pvalue=5.857257573144452e-10)


In [79]:
bad_tab_a1 = pd.crosstab(bad.A1, bad.A16)
print(bad_tab_a1)

A16    -
A1      
a    112
b    262


In [97]:
print('The Chi-Squared bad results for A1 and A16 crosstab: ', chisquare(bad_tab_a1, axis=None))

The Chi-Squared bad results for A1 and A16 crosstab:  Power_divergenceResult(statistic=60.160427807486634, pvalue=8.74324571943503e-15)


In [84]:
good_tab_a6 = pd.crosstab(good.A6, good.A16)
print(good_tab_a6)

A16   +
A6     
aa   19
c    62
cc   29
d     7
e    14
ff    7
i    14
j     3
k    14
m    16
q    51
r     2
w    33
x    32


In [96]:
print('The Chi-Squared good results for A6 and A16 crosstab: ', chisquare(good_tab_a6, axis=None))

The Chi-Squared good results for A6 and A16 crosstab:  Power_divergenceResult(statistic=192.0825082508251, pvalue=5.7469641750123415e-34)


In [86]:
bad_tab_a6 = pd.crosstab(bad.A6, bad.A16)
print(bad_tab_a6)

A16   -
A6     
aa   35
c    75
cc   12
d    23
e    11
ff   46
i    45
j     7
k    37
m    22
q    27
r     1
w    31
x     6


In [95]:
print('The Chi-Squared bad results for A6 and A16 crosstab: ', chisquare(bad_tab_a6, axis=None))

The Chi-Squared bad results for A6 and A16 crosstab:  Power_divergenceResult(statistic=192.88888888888889, pvalue=3.9285658300323e-34)


In [102]:
df_tab_a1 = pd.crosstab(df.A1, df.A16)
print(df_tab_a1)

A16    +    -
A1           
a     98  112
b    206  262


In [103]:
print('The Chi-Squared results for A1 and A16 crosstab: ', chisquare(df_tab_a1, axis=None))

The Chi-Squared results for A1 and A16 crosstab:  Power_divergenceResult(statistic=108.00589970501474, pvalue=2.947455715876286e-23)


In [104]:
df_tab_a4 = pd.crosstab(df.A4, df.A16)
print(df_tab_a4)

A16    +    -
A4           
l      2    0
u    256  263
y     45  118


In [105]:
print('The Chi-Squared results for A4 and A16 crosstab: ', chisquare(df_tab_a4, axis=None))

The Chi-Squared results for A4 and A16 crosstab:  Power_divergenceResult(statistic=637.561403508772, pvalue=1.545055409454076e-135)


In [107]:
df_tab_a6 = pd.crosstab(df.A6, df.A16)
print(df_tab_a6)

A16   +   -
A6         
aa   19  35
c    62  75
cc   29  12
d     7  23
e    14  11
ff    7  46
i    14  45
j     3   7
k    14  37
m    16  22
q    51  27
r     2   1
w    33  31
x    32   6


In [108]:
print('The Chi-Squared results for A6 and A16 crosstab: ', chisquare(df_tab_a6, axis=None))

The Chi-Squared results for A6 and A16 crosstab:  Power_divergenceResult(statistic=393.3201174743025, pvalue=1.1436746143425977e-66)


## Part 3 - Analysis and Interpretation

Now that you've looked at the data, answer the following questions:

- Interpret and explain the two t-tests you ran - what do they tell you about the relationships between the continuous features you selected and the class labels?
- Interpret and explain the two Chi-squared tests you ran - what do they tell you about the relationships between the categorical features you selected and the class labels?
- What was the most challenging part of this sprint challenge?

Answer with text, but feel free to intersperse example code/results or refer to it from earlier.

Our **T Test** is used to determine if the mean of our sample datasets are significantly different than the population means. The population means for A11 is 2.400 and A15 is 1017.386.

Tests:
T Test for good.A15  Ttest_1sampResult(statistic=2.3365846338986787, pvalue=0.020105944183966846)
T Test for bad.A15  Ttest_1sampResult(statistic=-23.858866910185004, pvalue=1.1004621903621446e-77)
Two tailed T Test for good.A15 and bad.A15  Ttest_indResult(statistic=4.680216020964486, pvalue=3.4520256956287944e-06)

T Test for good.A11  Ttest_1sampResult(statistic=6.115247963474785, pvalue=2.9376851337382315e-09)
T Test for bad.A11  Ttest_1sampResult(statistic=-18.211793670382843, pvalue=8.54839203623047e-54)
T Test for df.A11  Ttest_1sampResult(statistic=10.263115869755008, pvalue=4.298076068378268e-23)
Two tailed T Test for good.A11 and bad.A11  Ttest_indResult(statistic=11.667004222431277, pvalue=7.957718568079967e-29)

For A15, our mean in the good group was higher than the population group and the pvalue is low enough that we would reject our null hypothesis. The mean for the bad group is much lower than the mean for the population and with a high pvalue we would not be able to reject the null hypothesis. 

For A11, our good group again had a higher mean but with a high pvalue we can not reject the null hypothesis. For the bad group we see the mean is lower than the population and with a high pvalue we again can not reject the null hypothesis. 

Out two tailed t test shows that A15 and A11 play a larger role in determining creditworthiness for the good group than the bad group. 

My **Chi-Squared** results are as follows: 
Variables used: A16 is the constant variable between each.
A1, A4, A6 were the columns I used to match with A16. 
I split the data between the two possible results in A16, '+' was good or '-' was bad, into their own datasets. I then performed the Chi-Square test on each of these three columns:

The Chi-Squared good results for A1 and A16 crosstab:  Power_divergenceResult(statistic=38.36842105263158, pvalue=5.857257573144452e-10)
The Chi-Squared bad results for A1 and A16 crosstab:  Power_divergenceResult(statistic=60.160427807486634, pvalue=8.74324571943503e-15)

The Chi-Squared good results for A4 and A16 crosstab:  Power_divergenceResult(statistic=365.960396039604, pvalue=3.409649225976067e-80)
The Chi-Squared bad results for A4 and A16 crosstab:  Power_divergenceResult(statistic=55.183727034120736, pvalue=1.0977380519420046e-13)

The Chi-Squared good results for A6 and A16 crosstab:  Power_divergenceResult(statistic=192.0825082508251, pvalue=5.7469641750123415e-34)
The Chi-Squared bad results for A6 and A16 crosstab:  Power_divergenceResult(statistic=192.88888888888889, pvalue=3.9285658300323e-34)

The Chi-Squared results for the entire dataset for A1 and A16 crosstab:  Power_divergenceResult(statistic=108.00589970501474, pvalue=2.947455715876286e-23)
The Chi-Squared results for the entire dataset for A4 and A16 crosstab:  Power_divergenceResult(statistic=637.561403508772, pvalue=1.545055409454076e-135)
The Chi-Squared results for the entire dataset for A6 and A16 crosstab:  Power_divergenceResult(statistic=393.3201174743025, pvalue=1.1436746143425977e-66)

We use this Chi-Square to test if the variables are related. If they are related then we would expect a very low Chi-Square statistic and if they are unrelated we would get a large statistic. 
For our good datset here are the following statistics:
A1 and A16 - 38.368
A4 and A16 - 365.960
A6 and A16 - 192.083

For our bad dataset here are the following statistics:
A1 and A16 - 60.160
A4 and A16 - 55.184
A6 and A16 - 192.889

For the entire dataset, not splitting on A16, here is the following statistics:
A1 and A16 - 108.006
A4 and A16 - 637.561
A6 and A16 - 393.320

Looking at these Chi-Square statistics I would surmise that the A1 and A16 columns are closely related, at least more related than A4 or A6 is to A16. A1 would be more likely to help make inferences on A16 than A4 or A6 would. 

We can see that A1 in the good dataset appears to be more closely related to A16 then it is in the bad dataset, so if we are assuming that people in the good dataset are more creditworthy than those in the bad dataset, then A1 appears to be a good indicator of creditworthiness. 

A4 has a lower Chi-Square statistic in our bad dataset than in our good dataset, which I assume means that this is also a good determinant for creditworthiness, as it appears more frequently in the people who are not deemed worthy of additional credit. For those with bad credit, A4 plays a larger role in credit declines than it does for those with good credit. 

A6 has a near identical score between the good and bad dataset so it seems this is not a large factor in determining creditworthiness, as there is no change between the two groups for this variable. 

The most difficult  part of this challenge was determining which features were the best ones to use. It took some work to figure out which were the more extreme cases. But the whole thing was a bit scary. 